FRONTEND.md

```diff
@@ -0,0 +1,533 @@
+interpretation are new tracers living in the target scope.
+
+- The MyPy type of types is `Tuple[Tuple[AbstractValue,bool],...]`
+  [link](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/core.py#L1304)
+- The `bool` tuple elements represent “explicitness” of argument in a source Python program.
+  Explicit arguments explicitly appear in user-defined Python functions. Implicit arguments, in
+  contrast, are added by Jax to match Jaxpr typing requirements.
+
+The important properties of types are the following one. Abstract values found in types are allowed
+to include indices (called in Jax "de Brjuin indices" for some reason). The indices are described by
+objects of `DBIdx`, `InDBIdx` and `OutDBIdx` classes, all having their purpose.
+
+Example:
+
+```
+((DShapedArray(shape=(), dtype=int), False),
+ (DShapedArray(shape=(), dtype=int), False),
+ (DShapedArray(shape=[OutDBIdx(0),OutDBIdx(1),InDBIdx(0)], dtype=float), True)
+)
+```
+
+The above tuple may represent an output type of a Jaxpr program returning a 3D tensor along with the
+two of its three dynamic dimensions. The last dimension is assumed to be taken from the first input
+argument of the current program.
+
+### DBIdx
+
+`DBIdx(val)` are input type references. They are allowed in the shape values of `DShapedArray`
+objects, found in the *in_type* signature of Python/Jaxpr programs. The integer value of a
+reference are interpreted as an index in the current `in_type` signature tuple.
+
+Input type indices are:
+
+- Produced when calculating the input type of a nested program [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2321)
+- Consumed when creating argument tracers of a nested program [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2437)
+
+### InDBIdx/OutDBIdx
+
+`InDBIdx(val)/OutDBIdx(val)` are output type references. references allowed in the shape values of `DShapedArray`
+objects found in the *out_type* signatures of Python/Jaxpr programs.
+
+- *InDBIdx(val)* refers to the position in the `in_type` signature tuple of the Jaxpr/Python
+  program.
+- *OutDBIdx(val)* refers to the position in the `out_type` signature tuple of the Jaxpr/Python
+  program.
+
+Output type indices are:
+
+- Produced when calculating the output type of a nested program
+  [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2379)
+- Consumed when creating output tracers in the scope of an enclosing program
+  [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/pjit.py#L1333)
+
+### Primitives and binding
+
+Primitives are like functions in Jaxpr. If the primitive has a body, then binding is associated with
+tracing of the program representing primitive’s body. The binding of such a primitive includes:
```

Comment:
I think we briefly talked about this in another comment, but to my understanding _binding_ does not refer to the tracing itself, rather tracing is a _prerequisite_ of binding a primitive with a body.
Is binding then the process by which a primitive registers itself in the program representation? (if applicable, since there are contexts in which no program representation is generated --> maybe it would be more accurate to say the process by which a primitive applies itself to the interpreter stack?)

Reply:
(Aside:) If I think about the conclusion from this, do you know if that means that whenever an operation is invoked, the whole interpreter stack is immediately applied to it, rather than applying the stack once the whole program is captured? 🤔  (I guess this would apply in a jit context)

Reply:
I am not entirely sure what do you mean by applying a stack to a primitive, could you elaborate? I would say we apply a primitive to a stack and get an updated stack. Did you mean this?

Normally (in the classical case) we would pass a complete inner Jaxpr program to the primitive we want to bind, maybe as an Jax-IR data structure, but it would be a complete program already. That means that we have to perform the full tracing of the inner program (btw, we could benefit from caching here, I think. It might be helpful e.g. if we want to add several instances of the same loop). For the outer program, binding just creates new equation in its equation list.

For quantum tracing we do preserve all the stacks and record operations in an aside structure rather then bind immediately. But I though it is not directly related to the dynamic API problem, so I did not include these details into the document. Let me know, I can add this into the document as well.

I don't particulary like that we use different structures for quantum tracing and a different tracing method for classical tracing. I'd say that a better tracing library would still use only one queue structure in all modes (speaking of forking of Jax))

Reply:
> I am not entirely sure what do you mean by applying a stack to a primitive, could you elaborate? I would say we apply a primitive to a stack and get an updated stack. Did you mean this?

I was referencing the stack of interpreters that is setup by decorators like `grad` etc., which transform primitives in a certain way. So by "applying the interpreter stack to a primitive" I meant applying the processing function of each interpreter in the stack to the current primitive.

There was actually some great information shared about this on slack, so we can close this comment.